pkg/client/docker/kubeauth/cmd.go

```diff
@@ -0,0 +1,185 @@
+	PortFileStaleTime = 3 * time.Second
+)
+
+type authService struct {
+	portFile    string
+	kubeFlags   *genericclioptions.ConfigFlags
+	cancel      context.CancelFunc
+	configFiles []string
+}
+
+type PortFile struct {
+	Port       int    `json:"port"`
+	Kubeconfig string `json:"kubeconfig"`
+}
+
+func Command() *cobra.Command {
+	as := authService{kubeFlags: genericclioptions.NewConfigFlags(false)}
+	c := &cobra.Command{
+		Use:    CommandName,
+		Short:  "Launch Telepresence Kubernetes authenticator",
+		Args:   cobra.NoArgs,
+		Hidden: true,
+		RunE:   as.run,
+	}
+	flags := c.Flags()
+	flags.StringVar(&as.portFile, "portfile", "", "File where server existence is announced.")
+	as.kubeFlags.AddFlags(flags)
+	return c
+}
+
+func (as *authService) run(cmd *cobra.Command, _ []string) error {
+	ctx := cmd.Context()
+	cfg, err := client.LoadConfig(ctx)
+	if err != nil {
+		return err
+	}
+	ctx = client.WithConfig(ctx, cfg)
+
+	if as.portFile == "" {
+		return errcat.User.New("missing required flag --portfile")
+	}
+	grpcListener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return errcat.NoDaemonLogs.Newf("unable to open a port on localhost: %w", err)
+	}
+
+	ctx, err = logging.InitContext(ctx, "kubeauth", logging.RotateNever, false)
+	if err != nil {
+		return err
+	}
+	addr := grpcListener.Addr().(*net.TCPAddr)
+	dlog.Infof(ctx, "kubeauth listening on address %s", addr)
+
+	config := as.kubeFlags.ToRawKubeConfigLoader()
+	as.configFiles = config.ConfigAccess().GetLoadingPrecedence()
+	p := PortFile{
+		Port:       addr.Port,
+		Kubeconfig: strings.Join(as.configFiles, string(filepath.ListSeparator)),
+	}
+	pb, err := json.Marshal(&p)
+	if err != nil {
+		return err
+	}
+	if err = os.WriteFile(as.portFile, pb, 0o644); err != nil {
+		return err
+	}
+
+	ctx, as.cancel = context.WithCancel(ctx)
+	g := dgroup.NewGroup(ctx, dgroup.GroupConfig{})
+	g.Go("portfile-alive", as.keepPortFileAlive)
+	g.Go("portfile-watcher", as.watchFiles)
+	g.Go("grpc-server", func(ctx context.Context) error {
+		grpcHandler := grpc.NewServer()
+		authGrpc.RegisterAuthenticatorServer(grpcHandler, config)
+		sc := &dhttp.ServerConfig{Handler: grpcHandler}
+		return sc.Serve(ctx, grpcListener)
+	})
+	return g.Wait()
+}
+
+func (as *authService) keepPortFileAlive(ctx context.Context) error {
+	ticker := time.NewTicker(PortFileStaleTime)
+	defer func() {
+		ticker.Stop()
+		_ = os.Remove(as.portFile)
+		dlog.Debugf(ctx, "kubeauth removed %s", as.portFile)
+	}()
+	now := time.Now()
+	for {
+		if err := os.Chtimes(as.portFile, now, now); err != nil {
+			if !os.IsNotExist(err) {
+				return fmt.Errorf("failed to update timestamp on %s: %v", as.portFile, err)
+			}
+			// File is removed, so stop trying to update its timestamps and die
+			dlog.Info(ctx, "kubeauth exiting")
+			as.cancel()
+			return nil
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case now = <-ticker.C:
+		}
+	}
+}
+
+func (as *authService) watchFiles(ctx context.Context) error {
```

Comment:
I'm not sure why we have this, I assume it's to ensure that the kubeauth-foreground still uses the right kubeconfigs when we call a command ?

Reply:
This is necessary because if the user adds, deletes, or modifies one of the files listed in the `KUBECONFIG` path, then a restart is necessary. A restart is also necessary if the client detects that the config has changed and removes the file.